File: common/src/main/java/com/ultreon/mods/advanceddebug/api/common/IFormatter.java
package com.ultreon.mods.advanceddebug.api.common;

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext;

public interface IFormatter<T> {
    void format(T obj, IFormatterContext context);

    static <T extends IFormattable> IFormatter<T> create() {
        return IFormattable::format;
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/api/common/IFormatter.kt
package com.ultreon.mods.advanceddebug.api.common

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext

fun interface IFormatter<T> {
    fun format(obj: T, context: IFormatterContext)

    companion object {
        fun <T : IFormattable?> create(): IFormatter<T> {
            return IFormatter { obj: T, ctx: IFormatterContext ->
                obj?.format(
                    ctx
                ) ?: ctx.keyword("null")
            }
        }
    }
}